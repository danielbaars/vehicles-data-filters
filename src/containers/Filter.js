import React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import filterOptions from '../actions';
import { capFirstCharInString } from '../utility';

const Option = ({ option }) => {
  return (
    <option value={option}>{capFirstCharInString(option)}</option>
  );
};

export const FilterComponent = (props) => {
  const { keyValue, placeholder, options } = props;
  return [
    <h2 className='filters__title' key={`${keyValue}__title`}>{capFirstCharInString(keyValue)}:</h2>,
    <div className='filters__group input-group' key={`${keyValue}__select`}>
      <select onChange={event => props.filterOptions(keyValue, event.target.value)} name={`select-${keyValue}`} id={`select-${keyValue}`} value={options.length === 1 ? options[0] : (props[keyValue] || '')} className={`input-group-field filters__select filters__select--${keyValue}`}>
        <option value='' disabled>{placeholder}</option>
        {options.map((option) => {
          return (
            <Option option={option} key={`${keyValue}-${option}`} />
          );
        })}
      </select>
      <div className='input-group-button'>
        {props[keyValue]
          ? <button onClick={() => props.filterOptions(keyValue, undefined)} className='button small'>Clear</button>
          : <button className='button small hollow' disabled>Clear</button>
        }
      </div>
    </div>,
  ];
};

const mapStateToProps = (state) => {
  return state.filtersReducer;
};

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators({ filterOptions }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(FilterComponent);

Option.propTypes = {
  option: PropTypes.string.isRequired,
};

FilterComponent.propTypes = {
  keyValue: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  options: PropTypes.array.isRequired,
};
