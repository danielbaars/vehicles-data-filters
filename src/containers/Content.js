import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { uniqueArrayFromObjectValues } from '../utility';

import Row from '../components/Row';
import FramedBox from '../components/FramedBox';
import Results from '../components/Results';
import NoResults from '../components/NoResults';

import Filter from './Filter.js';

export class ContentComponent extends Component {
  getOptions(key) {
    const { data } = this.props;
    return uniqueArrayFromObjectValues(this.props[key] ? data : this.filterData(), key);
  }
  filterData() {
    const { data, colors, type, brand } = this.props;
    return data.filter(vehicle => [undefined, null, ...vehicle.colors].includes(colors) && [undefined, null, vehicle.type].includes(type) && [undefined, null, vehicle.brand].includes(brand));
  }
  render() {
    const { colors, type, brand } = this.props;
    const noFilters = [colors, type, brand].every(state => state === undefined);
    return (
      <div className="content__outer">
        <Row>
          <div className="small-12 medium-6 cell">
            <FramedBox className='filters'>
              <Filter keyValue='colors' placeholder='Select a color' options={this.getOptions('colors')} />
              <Filter keyValue='type' placeholder='Select a type' options={this.getOptions('type')} />
              <Filter keyValue='brand' placeholder='Select a brand' options={this.getOptions('brand')} />
            </FramedBox>
          </div>
          <div className="small-12 medium-6 cell">
            <FramedBox className='results'>
              {
                !noFilters
                  ? <Results results={this.filterData()} />
                  : <NoResults />
              }
            </FramedBox>
          </div>
        </Row>
      </div>
    );
  }
}

const mapStateToProps = state => state.filtersReducer;

export default connect(mapStateToProps)(ContentComponent);

ContentComponent.propTypes = {
  data: PropTypes.array.isRequired,
  colors: PropTypes.string,
  type: PropTypes.string,
  brand: PropTypes.string,
};

ContentComponent.defaultProps = {
  colors: undefined,
  type: undefined,
  brand: undefined,
};
