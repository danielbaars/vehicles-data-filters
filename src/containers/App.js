import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import '../style/foundation.css';
import '../style/app.css';

import Masthead from '../components/Masthead.js';
import Footer from '../components/Footer.js';
import Content from './Content.js';
import NoContent from '../components/NoContent.js';

export const AppComponent = ({ error, data, fetchData }) => {
  return (
    <div className="app">
      <Masthead />
      {
        error &&
          <NoContent fetchData={fetchData} />
      }
      {
        data &&
          <Content data={data} />
      }
      <Footer />
    </div>
  );
};

const mapStateToProps = (state) => {
  return {
    error: state.error,
    data: state.data,
  };
};

export default connect(mapStateToProps)(AppComponent);

AppComponent.propTypes = {
  error: PropTypes.bool.isRequired,
  data: PropTypes.array,
  fetchData: PropTypes.func.isRequired,
};

AppComponent.defaultProps = {
  data: [],
};
