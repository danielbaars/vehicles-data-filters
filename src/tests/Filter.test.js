import React from 'react';
import renderer from 'react-test-renderer';

import { FilterComponent } from '../containers/Filter';

describe('Filter', () => {
  const keyValue = 'colors';
  const placeholder = 'Select color';
  const options = ['red', 'green', 'blue'];
  const mockFilterOptions = jest.fn();
  const component = <FilterComponent keyValue={keyValue} placeholder={placeholder} options={options} filterOptions={mockFilterOptions} />;
  it('renders and matches our snapshot', () => {
    const rComponent = renderer.create(component);
    const tree = rComponent.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
