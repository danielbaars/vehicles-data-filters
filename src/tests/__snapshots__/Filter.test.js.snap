// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter renders and matches our snapshot 1`] = `
Array [
  <h2
    className="filters__title"
>
    Colors
    :
</h2>,
  <div
    className="filters__group input-group"
>
    <select
        className="input-group-field filters__select filters__select--colors"
        id="select-colors"
        name="select-colors"
        onChange={[Function]}
        value=""
    >
        <option
            disabled={true}
            value=""
        >
            Select color
        </option>
        <option
            value="red"
        >
            Red
        </option>
        <option
            value="green"
        >
            Green
        </option>
        <option
            value="blue"
        >
            Blue
        </option>
    </select>
    <div
        className="input-group-button"
    >
        <button
            className="button small hollow"
            disabled={true}
        >
            Clear
        </button>
    </div>
</div>,
]
`;
